src/lib/rtl/rtl_time_estimator.cpp

```diff
@@ -0,0 +1,267 @@
+		// Use actual time estimate to compute the safer time estimate with additional scale factor and a margin
+		time_estimate.safe_time_estimate = _param_rtl_time_factor.get() * _time_estimate + _param_rtl_time_margin.get();
+
+	} else {
+		time_estimate.valid = false;
+	}
+
+	time_estimate.timestamp = hrt_absolute_time();
+	return time_estimate;
+}
+
+void RtlTimeEstimator::update()
+{
+	_vehicle_status_sub.update();
+	_wind_sub.update();
+
+	if (_parameter_update_sub.updated()) {
+		parameter_update_s param_update;
+		_parameter_update_sub.copy(&param_update);
+
+		// If any parameter updated, call updateParams() to check if
+		// this class attributes need updating (and do so).
+		updateParams();
+	}
+}
+
+void RtlTimeEstimator::addVertDistance(float alt)
+{
+	if (PX4_ISFINITE(alt)) {
+		_is_valid = true;
+
+		_time_estimate += calcVertTimeEstimate(alt);
+	}
+}
+
+void RtlTimeEstimator::addDistance(float hor_dist, const matrix::Vector2f &direction, float vert_dist)
+{
+	if (PX4_ISFINITE(hor_dist) && PX4_ISFINITE(vert_dist)) {
+		_is_valid = true;
+
+		float hor_time_estimate{0.f};
+
+		if (hor_dist > FLT_EPSILON && PX4_ISFINITE(hor_dist)) {
+			hor_time_estimate = hor_dist / getCruiseGroundSpeed(direction.normalized());
+		}
+
+		float ver_time_estimate{calcVertTimeEstimate(vert_dist)};
+
+		_time_estimate += math::max(hor_time_estimate, ver_time_estimate);
+
+	}
+}
+
+void RtlTimeEstimator::addWait(float time_s)
+{
+	if (PX4_ISFINITE(time_s)) {
+		_is_valid = true;
+
+		if (time_s > FLT_EPSILON) {
+			_time_estimate += time_s;
+		}
+	}
+}
+
+void RtlTimeEstimator::addDescendMCLand(float alt)
+{
+	if (PX4_ISFINITE(alt)) {
+		_is_valid = true;
+
+		if (alt < -FLT_EPSILON && PX4_ISFINITE(alt)) {
+			_time_estimate += -alt / getHoverLandSpeed();
+		}
+	}
+}
+
+float RtlTimeEstimator::getCruiseGroundSpeed(const matrix::Vector2f &direction_norm)
+{
+	float cruise_speed = getCruiseSpeed();
+
+	if (_vehicle_status_sub.get().vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING) {
+		matrix::Vector2f wind = get_wind();
+
+		const float wind_along_dir = wind.dot(direction_norm);
+		const float wind_across_dir = matrix::Vector2f(wind - direction_norm * wind_along_dir).norm();
+
+		// Note: use fminf so that we don't _rely_ on tailwind towards direction to make RTL more efficient
+		const float ground_speed = sqrtf(cruise_speed * cruise_speed - wind_across_dir * wind_across_dir) + fminf(
+						   0.f, wind_along_dir);
+
+		cruise_speed = ground_speed;
+	}
+
+	return cruise_speed;
+}
+
+float RtlTimeEstimator::calcVertTimeEstimate(float alt)
+{
+	float vertical_rate{0.1f};
+	float time_estimate{0.f};
+
+	if (alt > FLT_EPSILON) {
+		vertical_rate = getClimbRate();
+
+	} else {
+		vertical_rate = getDescendRate();
+	}
+
+	float abs_alt = fabsf(alt);
+
+	if (abs_alt > FLT_EPSILON) {
+		time_estimate = abs_alt / vertical_rate;
+	}
+
+	return time_estimate;
+}
+
+float RtlTimeEstimator::getCruiseSpeed()
+{
+	float ret = 1e6f;
```

Comment:
Ideally we would invalidate the estimate if one of the params cannot be fetched.

Reply:
This is just copied code, but i think the idea was that for vtol types, where it does not matter, just add a very high number so it doesn't add to the time estimate. But that part of code should anyways be rearranged so it can properly estimate for different vtol types.

Reply:
can be changed later. Better invalidate the estimate instead of publishing a small one.